### PR TITLE
CASMINST-3671: Add some quality of life changes to the xnames package

### PR DESCRIPTION
**Pull request by rsjostrand-hpe**
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

- Added the `GenericXname` interface to the `xnames` package. It is meant to provide a generic way to store or pass a xnames structure, such as to a function that can handle multiple different xnames types.

  All of the xnames structures now implement the following functions:
  - `String() string` - build the xname string for the given structure
  - `Type() xnametypes.HMSType` - get the Type enumeration of the structure.
  - `ParentGeneric() GenericXname` - get the xname 
  - `IsController() bool` - returns true if the structure is a controller type, i.e. that  would host a Redfish entry point

- Corrected the name of the `xnames` package, as it was originally set to `xname`.
- Changed the `xnames.FromString` method to use the GenericInterface type. There exists no consumer of the FromString function so this should not cause any breakages. 

_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Future work required by [https://jira-pro.its.hpecorp.net:8443/browse/CASMINST-3671](CASMINST-3671)

## Testing

_List the environments in which these changes were tested._

Tested on:
  * Local development environment

Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)? N/A
- Were continuous integration tests run? If not, why? Yes, unit tests ran.
- Was upgrade tested? If not, why? N/A
- Was downgrade tested? If not, why? N/A

Ran updated unit tests with my changes. 

The changes within this PR are currently being used by my WIP changes to cray-site-init.

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_
Low risk.

## Pull Request Checklist

- [X] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [X] CHANGELOG.md updated
- [X] Testing is appropriate and complete, if applicable

### CODE REVIEW

File: xnames/generator/types.go.tpl [hunk 1]
```diff
@@ -36,6 +36,11 @@ type {{ $xnameType.Entry.Type }} struct {
 {{- end }}
 }
 
+// Type will return the corresponding HMSType
```
Comment (nieuwsma) [inline xnames/generator/types.go.tpl:39]: can you tell me more about .go.tpl files? im not familiar with them.

**Comment by nieuwsma** [comment]
I dont have any issues with this. Can I get a walk through tomorrow?

File: xnames/generator/types.go.tpl (hunk 1 above)
Comment (rsjostrand-hpe) [reply xnames/generator/types.go.tpl:39]: Here is some more context: https://github.com/Cray-HPE/hms-xname#code-generation

The .go.tpl files are template files that the generator uses to generate the real types.go and util.go. It uses the golang template language: https://pkg.go.dev/text/template

**Review by nieuwsma — APPROVED** [review]
Ryan and I talked about the changes he needs to make. I approve of the direction of it.

**Review by shunr-hpe — APPROVED** [review]

**Pull request merged**